### PR TITLE
Find resident deduplication fix

### PR DESCRIPTION
**Pull request by Duslerke**
# What
 - When first name, last name or both are missing the code won't proceed to Uprn & Dob matching rules.
 - All 3 of the Date of birth parts (Year, Month, Day) are required for the Dob rule to flag resident as duplicate. (Previously it was 1+)
 - More extensive check for empty values. Previously it was just a "null" check, now it also checks for empty and whitespace string as well.
 - Added NHS number rule for flagging resident as duplicate.
 - Minor code tweaks to make it more efficient & readable.

# Why
 - Deduplication isn't working as expected. It's flagging some residents as duplicate, when they are not. And it's not flagging some as duplicate when they are.

# Screenshots
 - No screenshot

# Link to ticket
 - Trello card: [107](https://trello.com/c/wkcGQa7x/107-bug-de-duping-on-api-too-aggressive).

# Notes
 - Suppressed the Globalisation suggestion as we're not building an international application that would expect multi-cultural strings. Did this to remove that obnoxious green squiggly line that made code harder to read. Judging from how long these warnings stayed around, I assume everyone's of the same opinion that adding those "Ignore Culture" options to every string method like .ToUpper() would make the code unpleasant and harder to read due to all the extra distraction from the key detail.
 - Require both first and last names to be non-empty, like mentioned in the commit. Having last name & uprn, or first name & dob is quite clearly not enough information to guarantee that the resident in question is a duplicate.
 - Added more extensive empty checks because, similarly as above, we can't guarantee that a person with a given first name, last name and empty uprn is a duplicate (even when empty uprns math: " " == " ").
 - All 3 parts of the date required as, like above, having a first name, last name and DobDay is not enough to guarantee resident is a duplicate. Same with other variations of Dob fields - at best (if DobYear & DobDay are provided) you can expect 12 combinations of dates containing the provided information, which is not great.
 - Any resident can have 1+ of NHS numbers, however no resident can have the NHS number that other has. Hence finding NHS number match is enough to guarantee resident is a duplicate. (Read more in commit message)

### CODE REVIEW

**Comment by Duslerke** [comment]
Damn it, I forgot to add NHS number trimming... not sure if this is an issue or not as I haven't seen the data. Will add it and the test now.

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 1]
```diff
@@ -5,89 +5,477 @@
...
+            };
+
+            DatabaseContext.ResidentEntities.Add(existingResidentNoNhsNumberMatch);
+            DatabaseContext.ResidentEntities.Add(existingResidentNhsNumberMatch);
             DatabaseContext.SaveChanges();
-            var response = _classUnderTest.FindResident(findResidentCommand);
-            var createdRecord = DatabaseContext.ResidentEntities.Find(response);
-            var expectedRecord = new ResidentEntity
+
+            var nhsNumberMatchResidentId = DatabaseContext.ResidentEntities.FirstOrDefault(r => r.NhsNumber == matchingNhsNumber)?.Id;
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:59]: Do we need this variable? Could we just match against ```existingResidentNhsNumberMatch.Id``` ? That way we don't rely on the database on the both sides of the assertion

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 2]
```diff
@@ -5,89 +5,477 @@
...
+            var existingResidentUprnMatch = new ResidentEntity
+            {
+                NhsNumber = _faker.Random.Hash(),
+                Uprn = matchingUprn,
+                FirstName = matchingFirstName,
+                LastName = matchingLastName
+            };
+
+            // testing the no match path (adding this 1 line, so I wouldn't need to copy the whole test)
+            if (matchingEmptyNhsNumber == "nhs number") existingResidentEmptyNhsNumberMatch.NhsNumber = _faker.Random.Hash();
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:112]: This sounds like a separate test case.
Is this a valid way to do it ```[TestCase(_faker.Random.Hash())]```
Also, seems like ```matchingEmptyNhsNumber``` isn't always empty, is there something else we could name it, if you choose to keep it as a single test?

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 3]
```diff
@@ -5,89 +5,477 @@
...
+
+            DatabaseContext.ResidentEntities.Add(existingResidentEmptyNhsNumberMatch);
+            DatabaseContext.ResidentEntities.Add(existingResidentUprnMatch);
+            DatabaseContext.SaveChanges();
+
+            var uprnMatchResidentId = DatabaseContext.ResidentEntities.FirstOrDefault(r => r.Uprn == matchingUprn)?.Id;
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:118]: Same as before, can't we access the Id through ```existingResidentUprnMatch.Id``` ?

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 4]
```diff
@@ -5,89 +5,477 @@
...
+            var existingResidentEmptyUprnMatch = new ResidentEntity
+            {
+                NhsNumber = _faker.Random.Hash(),
+                Uprn = matchingEmptyUprn, // matching empty uprn
+                FirstName = matchingFirstName,
+                LastName = matchingLastName
+            };
+
+            // testing a case path, where uprn is not empty, but still does not match
+            if (matchingEmptyUprn == "uprn") existingResidentEmptyUprnMatch.Uprn = _faker.Random.Hash();
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:172]: Same as the test above

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 5]
```diff
@@ -5,89 +5,477 @@
...
+                Uprn = emptyMatchingUprn, // check to see if we fall through to the next rule on empty match
+                FirstName = matchingFirstName,
+                LastName = matchingLastName,
+                DobDay = matchingDOBDay,
+                DobMonth = matchingDOBMonth,
+                DobYear = matchingDOBYear
+            };
+
+            // checking rule falling through to the next on Uprn no match (not just empty)
+            if (emptyMatchingUprn == "uprn") existingResidentDOBMatch.Uprn = _faker.Random.Hash();
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:243]: Same as above

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 6]
```diff
@@ -5,89 +5,477 @@
...
+            DatabaseContext.ResidentEntities.Add(existingResidentNoDOBMatch);
+            DatabaseContext.ResidentEntities.Add(existingResidentDOBMatch);
+            DatabaseContext.SaveChanges();
+
+            var dobMatchResidentId = DatabaseContext.ResidentEntities
+                .FirstOrDefault(r =>
+                    r.DobDay == matchingDOBDay &&
+                    r.DobMonth == matchingDOBMonth &&
+                    r.DobYear == matchingDOBYear
+                )?.Id;
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:254]: Same as above

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 7]
```diff
@@ -5,89 +5,477 @@
...
+                DobYear = emptyMatchingDOBYear
+            };
+
+            // checking a non-empty match case
+            if (emptyMatchingDOBYear == "1994")
+            {
+                existingResidentEmptyDOBMatch.DobDay = _faker.Random.Hash();
+                existingResidentEmptyDOBMatch.DobMonth = _faker.Random.Hash();
+                existingResidentEmptyDOBMatch.DobYear = _faker.Random.Hash();
+            }
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:329]: Same here

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs [hunk 8]
```diff
@@ -5,89 +5,477 @@
...
+            DatabaseContext.ResidentEntities.Add(existingResidentWithNormalNhsNumber);
+            DatabaseContext.ResidentEntities.Add(existingResidentWithNonTrimmedNhsNumber);
+            DatabaseContext.SaveChanges();
+
+            var ResidentWithNormalNhsNumberId1 = DatabaseContext.ResidentEntities
+                .FirstOrDefault(r => r.NhsNumber == nhsNumber1)?.Id;
+
+            var ResidentWithNormalNhsNumberId2 = DatabaseContext.ResidentEntities
+                .FirstOrDefault(r => r.NhsNumber == nonTrimmedNHSNumber2)?.Id;
+
```
Comment (kosiakkatrina) [inline cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:461]: Can it be ```existingResidentWithNormalNhsNumber.Id``` and ```existingResidentWithNonTrimmedNhsNumber.Id``` ?

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 1 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:59]: Unfortunately no. The Ids are auto-generated upon inserting a record, meaning we don't know what the id is until we insert the object, and then retrieve it to look at what the id is.

File: cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+namespace cv19ResSupportV3.V4.Helpers
+{
+    public static class Predicates
+    {
+        public static bool IsNotNullAndNotEmpty(string testString)
+        {
+            return testString != null && testString.Trim() != "";
+        }
+    }
+}
```
Comment (kosiakkatrina) [inline cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs:10]: Do we need to define this separately?
We could just use the ```String.IsNullOrEmpty(s)``` method whenever we want to check if string is null or empty

File: cv19ResSupportV3/V3/Gateways/ResidentGateway.cs [hunk 10]
```diff
@@ -164,37 +165,55 @@ public Resident PatchResident(int id, PatchResident command)
...
-                        command.FirstName.ToUpper().Trim() &&
-                        x.LastName.ToUpper()
-                            .Trim() ==
-                        command.LastName.ToUpper().Trim());
-                    if (response != null)
-                    {
-                        return response.Id;
-                    }
+                    var matchingResident = _helpRequestsContext.ResidentEntities
+                        .FirstOrDefault(r => r.NhsNumber.Trim().ToUpper() == command.NhsNumber.Trim().ToUpper());
```
Comment (kosiakkatrina) [inline cv19ResSupportV3/V3/Gateways/ResidentGateway.cs:175]: Can NhsNumber contain letters? Not sure if we need ```ToUpper``` here

File: cv19ResSupportV3/V3/Gateways/ResidentGateway.cs [hunk 11]
```diff
@@ -164,37 +165,55 @@ public Resident PatchResident(int id, PatchResident command)
...
+
+                    if (matchingResident != null)
+                        return matchingResident.Id;
                 }
 
-                if (command.DobDay != null || command.DobMonth != null || command.DobYear != null)
+                // If Fname or Lname are missing, no point checking Uprn or Dob. When these two fields
+                // are missing, there's no way to confirm whether it's the same person or not (unless nhs numbers match).
+                if (Predicates.IsNotNullAndNotEmpty(command.FirstName) &&
+                    Predicates.IsNotNullAndNotEmpty(command.LastName))
```
Comment (kosiakkatrina) [inline cv19ResSupportV3/V3/Gateways/ResidentGateway.cs:184]: Nice!

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 2 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:112]: Yeah the naming of this _("matchingEmptyNhsNumber")_ variable is a tricky business 😂 Will try to think of something better before merging.

I don't think that case deserves a separate test as it falls under the same edge case category as the others, which is that an NHS number de-duplication rule should find no match and fall through to the next one when it's either empty or simply does not match _(Essentially NHS rule should fall through to the next one on these 4 conditions - hence TestCases)_. An alternative is to copy paste this entire test for that one Test Case, which doesn't feel appropriate in terms of tests readability when this TestCase is the same category as the other ones.

Unfortunately, `[TestCase(_faker.Random.Hash())]` won't work, because the decorator attribute expects a concrete value. So dropping in a function call or an expression will give an error.

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 3 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:118]: Unfortunately no, it's because we auto-generate Ids, so we don't know what it will be unless we check that. Due to NUnit tests running in a different order than shown (and maybe even asynchronously), it would be hard to reason about the Id that will be given. Even if we guess it, and the hardcode it - any time we add new tests, there's a possibility that our hardcoded guess could fail, due to the order changing order of how the tests run. _(Our tests setup doesn't clear DB properly, but honestly, even if it did, I feel relying on a DB query that's tested Microsoft isn't as bad as hardcoding a guess.)_

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 4 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:172]: Same answer as for the first 2 cases of this 😁

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 5 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:243]: Same answer as for the first 2 cases of this 😁

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 6 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:254]: Same answer as for the first 2 cases of this 😁

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 8 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs:461]: Same answer as for the first 2 cases of this 😁 

We have 2 routes here:
1) Guess an Id and hardcode it.
2) Use ".FirstOrDefault()" Linq expression, which is already tested by Microsoft, and doesn't impact what's being tested.

I feel the later is more elegant and safer solution.

File: cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs (hunk 9 above)
Comment (Duslerke) [reply cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs:10]: I remember reading something about the IsNullOrEmpty before, and people were suggesting not to use it as it depends on your environment (essentially side effects). So I wrote a custom one, which also ads the flexibility of extending the definition of empty _(maybe add check for unicode whitespaces or something of the sort)_ 😁

File: cv19ResSupportV3/V3/Gateways/ResidentGateway.cs (hunk 10 above)
Comment (Duslerke) [reply cv19ResSupportV3/V3/Gateways/ResidentGateway.cs:175]: Yeah, at the time of writing the code I didn't know that myself. Only figured that out after resolving a merge conflict and forgot to remove it 😁 Will do now.

**Review by kosiakkatrina — APPROVED** [review]

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 3 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs]: Nvm... adding a record to DB has a side effect that updates the object that you're inserting. Good shout!

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 6 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs]: Nvm... adding a record to DB has a side effect that updates the object that you're inserting. Good shout!

File: cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs (hunk 8 above)
Comment (Duslerke) [reply cv19ResSupportV3.Tests/V3/Gateways/ResidentGatewayTest.cs]: Nvm... adding a record to DB has a side effect that updates the object that you're inserting. Good shout!

File: cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs (hunk 9 above)
Comment (Duslerke) [reply cv19ResSupportV3/V4/Helpers/PredicateHelpers.cs:10]: Agreed to leave this for later, as this needs looking into.

File: cv19ResSupportV3/V3/Gateways/ResidentGateway.cs (hunk 10 above)
Comment (Duslerke) [reply cv19ResSupportV3/V3/Gateways/ResidentGateway.cs]: Done.
